### PR TITLE
Gr01.6

### PR DESCRIPTION
**Pull request by TylerTMizuyabu**
# AWS CAC Solution PR Template

## Pull Request Type
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation Update
- [ ] Other (Please specify)

## Description

Adds a lambda for guardrail check 1.06. This lambda queries Identity Center and IAM for all users, then checks their policies. It gives compliant if there is at least one privilegeduser with admin access, and if no non-privileged users have admin access.

Also updates 02 because I believe the backslash isn't needed to escape the '{' character

## Related Issue(s)
List any related issues here. Include `Fixes #issue_number` or `Closes #issue_number` if applicable.

## How Has This Been Tested?
Describe how you have tested these changes. Include details of the testing environment and the tests you ran to verify the functionality.

### CloudFormation Linter Results
Include the output of the CloudFormation linter (if used).

### Deployment Test Results
Provide details of any deployment tests you have conducted.

## Checklist:
- [ ] My code follows the code style of this project.
- [ ] I have updated any related documentation.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] My changes generate no new warnings.
- [ ] I have added comments to my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes require a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have checked that the final deliverable aligns with the CloudFormation best practices.

## Additional Notes
Include any additional information that you think is important for reviewers.

### CODE REVIEW

File: src/lambda/gc01_check_dedicated_admin_account/app.py [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+            atleast_one_priveleged_user_has_admin_access = policies_grant_admin_access(managed_policies, inline_policies)
+            
+        elif user_name in non_privileged_user_list:
+            managed_policies = fetch_aws_managed_user_policies(user_name)
+            inline_policies = fetch_inline_user_policies(user_name)
+            if policies_grant_admin_access(managed_policies, inline_policies):
+                non_privileged_user_with_admin_access.append(user_name)
+                
+    for instance_arn in sso_users_by_instance.keys():
+        admin_permssion_sets_by_account=get_admin_permission_sets_for_instance_by_account(instance_arn)
```
Comment (david-dearden-ssc) [inline src/lambda/gc01_check_dedicated_admin_account/app.py:492]:
```suggestion
        admin_perimssion_sets_by_account=get_admin_permission_sets_for_instance_by_account(instance_arn)
```

**Review by david-dearden-ssc — CHANGES_REQUESTED** [review]

File: src/lambda/gc01_check_dedicated_admin_account/app.py [hunk 2]
```diff
@@ -0,0 +1,632 @@
...
+def fetch_sso_users():
+    instances = fetch_sso_instances()
+    users_by_instance = {}
+    for i in instances:
+        if i.get("Status") != "Active":
+            continue
+        
+        instance_id = i.get("IdentityStoreId")
+        instance_arn = i.get("InstanceArn")
+        users_by_instance[instance_id]
```
[the remark this answers is not in the document]
Comment (TylerTMizuyabu) [reply src/lambda/gc01_check_dedicated_admin_account/app.py:208]: I think this line was leftover from some debuging. doesn't blow up. I have the function running on lza

File: src/lambda/gc01_check_dedicated_admin_account/app.py [hunk 3]
```diff
@@ -0,0 +1,632 @@
...
+                    policies = policies + response.get("AttachedPolicies", [])
+                    inner_marker = response.get("Marker")
+                    if not inner_marker:
+                        break
+            marker = response.get("Marker")
+            
+            if not marker:
+                break
+            
+        return policies
```
[the remark this answers is not in the document]
Comment (TylerTMizuyabu) [reply src/lambda/gc01_check_dedicated_admin_account/app.py:315]: Inline policies we need to actually look at their policy doc statements so we have to fetch that information. The managed one we don't and also I don't think we can. At least I haven't been able to find an AWS api that does this for managed policies

[the remark this answers is not in the document]
Comment (david-dearden-ssc) [reply src/lambda/gc01_check_dedicated_admin_account/app.py:315]: Ah, ok, makes sense, I guess `policies_grant_admin_access` has a fallback for a missing policy document.

File: src/lambda/gc01_check_dedicated_admin_account/app.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TylerTMizuyabu) [reply src/lambda/gc01_check_dedicated_admin_account/app.py:208]: actually I meant to initialize the list here so the next comment wouldn't be an issue

File: src/lambda/gc01_check_dedicated_admin_account/app.py [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+            continue
+        
+        instance_id = i.get("IdentityStoreId")
+        instance_arn = i.get("InstanceArn")
+        users_by_instance[instance_id]
+        nextToken = None
+        try:
+            while True:
+                response = AWS_IDENTITY_STORE_CLIENT.list_users(IdentityStoreId=instance_id, NextToken=nextToken) if nextToken else AWS_IDENTITY_STORE_CLIENT.list_users(IdentityStoreId=instance_id)
+                users_by_instance[instance_arn] = users_by_instance[instance_arn] + response.get("Users", [])
```
[the remark this answers is not in the document]
Comment (TylerTMizuyabu) [reply src/lambda/gc01_check_dedicated_admin_account/app.py:214]: response to previous comment fixes this

File: src/lambda/gc01_check_dedicated_admin_account/app.py [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+def fetch_sso_users():
+    instances = fetch_sso_instances()
+    users_by_instance = {}
+    for i in instances:
+        if i.get("Status") != "Active":
+            continue
+        
+        instance_id = i.get("IdentityStoreId")
+        instance_arn = i.get("InstanceArn")
+        users_by_instance[instance_id] = []
```
Comment (david-dearden-ssc) [inline src/lambda/gc01_check_dedicated_admin_account/app.py:209]:
```suggestion
        users_by_instance[instance_arn] = []
```

**Review by david-dearden-ssc — CHANGES_REQUESTED** [review]

File: src/lambda/gc01_check_dedicated_admin_account/app.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TylerTMizuyabu) [reply src/lambda/gc01_check_dedicated_admin_account/app.py:316]: yeah it takes in managed and inline policies as different parameters right. So it treats them differently. for managed policies we are just looking at the name and no policy doc. We actually don't expect the objects in that list to have a policy doc. For inline policies we expect the policy doc to be there

File: src/lambda/gc01_check_dedicated_admin_account/app.py [hunk 6]
```diff
@@ -0,0 +1,632 @@
...
+    except botocore.exceptions.ClientError as ex:
+        ex.response["Error"]["Message"] = "InternalError"
+        ex.response["Error"]["Code"] = "InternalError"
+        raise ex
+    
+    return permission_sets
+  
+def user_assigned_to_permission_set(user_id, instance_arn, account_id, permission_set_arn):
+    try:
+        response = AWS_SSO_ADMIN_CLIENT.list_account_assignments(AccountId=account_id, InstanceArn=instance_arn, PermissionSetArn=permission_set_arn)
```
[the remark this answers is not in the document]
Comment (TylerTMizuyabu) [reply src/lambda/gc01_check_dedicated_admin_account/app.py]: yup, updated to loop

**Review by david-dearden-ssc — APPROVED** [review]

**Review by a-shokri-ssc — APPROVED** [review]

**Pull request merged**